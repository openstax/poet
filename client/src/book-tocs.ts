import { EventEmitter, type TreeItem, TreeItemCollapsibleState, Uri, type TreeDataProvider } from 'vscode'

import { type BookToc, type ClientTocNode, BookRootNode, TocNodeKind, type ClientPageish } from '../../common/src/toc'
import { TocItemIcon } from './toc-trees-provider'

export type BookOrTocNode = BookToc | ClientTocNode

const toClientTocNode = (n: ClientPageish): ClientTocNode => ({ type: TocNodeKind.Page, value: n })

export class TocsTreeProvider implements TreeDataProvider<BookOrTocNode> {
  private readonly _onDidChangeTreeData = new EventEmitter<void>()
  readonly onDidChangeTreeData = this._onDidChangeTreeData.event

  public includeFileIdsForFilter = false
  private bookTocs: BookToc[]
  private orphans: BookOrTocNode[]
  private readonly parentsMap = new Map<BookOrTocNode, BookOrTocNode>()

  constructor() {
    this.bookTocs = []
    this.orphans = []
  }

  public toggleFilterMode() {
    this.includeFileIdsForFilter = !this.includeFileIdsForFilter
    this._onDidChangeTreeData.fire()
  }

  public update(n: BookToc[], o: ClientPageish[]) {
    this.bookTocs = n
    this.parentsMap.clear()
    this.bookTocs.forEach(n => { this.recAddParent(n) })
    this.orphans = o.map(toClientTocNode)
    this._onDidChangeTreeData.fire()
  }

  private recAddParent(node: BookOrTocNode) {
    const kids = this.getChildren(node)
    kids.forEach(k => {
      this.parentsMap.set(k, node)
      this.recAddParent(k)
    })
  }

  public getTreeItem(node: BookOrTocNode): TreeItem {
    if (node.type === BookRootNode.Singleton) {
      const uri = Uri.parse(node.absPath)
      return {
        iconPath: TocItemIcon.Book,
        collapsibleState: TreeItemCollapsibleState.Collapsed,
        label: node.title,
        description: node.slug,
        resourceUri: uri,
        command: { title: 'open', command: 'vscode.open', arguments: [uri] }
      }
    } else if (node.type === TocNodeKind.Page) {
      const uri = Uri.parse(node.value.absPath)
      const title = node.value.title ?? 'Loading...'
      const ret = this.includeFileIdsForFilter ? { label: `${title} (${node.value.fileId})` } : { label: title, description: node.value.fileId }
      return {
        ...ret,
        iconPath: TocItemIcon.Page,
        collapsibleState: TreeItemCollapsibleState.None,
        resourceUri: uri,
        command: { title: 'open', command: 'vscode.open', arguments: [uri] },
        contextValue: TocNodeKind.Page
      }
    } else {
      return {
        iconPath: TocItemIcon.Subbook,
        collapsibleState: TreeItemCollapsibleState.Collapsed,
        label: node.value.title,
        contextValue: TocNodeKind.Subbook
      }
    }
  }

  public getChildren(node?: BookOrTocNode) {
    let kids: BookOrTocNode[] = []
    if (node === undefined) {
      return [...this.bookTocs, ...this.orphans]
    } else if (node.type === BookRootNode.Singleton) {
      kids = node.tocTree
    } else if (node.type === TocNodeKind.Page) {
      kids = []
    } else {
      kids = node.children
    }
    return kids
  }

  public getParent(node: BookOrTocNode) {
    return this.parentsMap.get(node)
  }

  public getParentBook(node: BookOrTocNode): BookToc | undefined {
    let bookNode: BookOrTocNode | undefined = node
    // Books do not have parent books
    if (bookNode.type === BookRootNode.Singleton) return undefined
    while (bookNode !== undefined && bookNode.type !== BookRootNode.Singleton) {
      bookNode = this.getParent(bookNode)
    }
    return bookNode?.type === BookRootNode.Singleton ? bookNode : undefined
  }

  public getParentBookIndex(node: BookOrTocNode) {
    const parentBook = this.getParentBook(node)
    let idx: number | undefined
    if (parentBook !== undefined) {
      idx = this.bookTocs.findIndex((b) => b.absPath === parentBook.absPath)
    }
    return idx === -1 ? undefined : idx
  }
}
