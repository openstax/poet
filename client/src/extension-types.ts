/* istanbul ignore file */
export enum PanelType {
  TOC_EDITOR = 'openstax.tocEditor',
  IMAGE_MANAGER = 'openstax.imageManager',
  CNXML_PREVIEW = 'openstax.cnxmlPreview'
}
export enum OpenstaxCommand {
  SHOW_TOC_EDITOR = 'openstax.showTocEditor',
  SHOW_CNXML_PREVIEW = 'openstax.showPreviewToSide',
  SHOW_IMAGE_MANAGER = 'openstax.showImageManager'
}
