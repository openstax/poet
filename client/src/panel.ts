import vscode from 'vscode'
import { LanguageClient } from 'vscode-languageclient/node'

// Modified from https://github.com/microsoft/vscode/blob/main/extensions/markdown-language-features/src/util/dispose.ts
class Disposer implements Disposable {
  private _disposed = false
  private registeredDisposables: vscode.Disposable[] = [] // TODO: change type to Disposable[]
  private onDidDisposeListeners: Array<() => void> = []

  onDidDispose(listener: () => void): void {
    this.onDidDisposeListeners.push(listener)
  }

  private disposeAll(): void {
    for (const disposable of this.registeredDisposables) {
      console.log(disposable)
      disposable.dispose()
    }
    this.registeredDisposables = []
  }

  dispose(): void {
    if (this._disposed) {
      return
    }
    this._disposed = true
    for (const listener of this.onDidDisposeListeners) {
      listener()
    }
    this.onDidDisposeListeners = []
    this.disposeAll()
  }

  registerDisposable<T extends vscode.Disposable>(value: T): T {
    if (this._disposed) {
      value.dispose()
    } else {
      this.registeredDisposables.push(value)
    }
    return value
  }

  disposed(): boolean {
    return this._disposed
  }
}

export abstract class Panel<InMessage, OutMessage> implements Disposable, Messageable<InMessage, OutMessage> {
  protected readonly panel: vscode.WebviewPanel
  private readonly disposer: Disposable

  constructor(initPanel: () => vscode.WebviewPanel) {
    this.disposer = new Disposer()
    this.panel = initPanel()
    this.registerDisposable(this.panel)
    this.panel.onDidDispose(() => this.dispose())
    this.panel.webview.onDidReceiveMessage((message) => {
      this.handleMessage(message).catch((err: Error) => {
        void vscode.window.showErrorMessage(err.message)
        throw err
      })
    })
  }

  abstract handleMessage(message: InMessage): Promise<void>
  async postMessage(message: OutMessage): Promise<void> {
    await this.panel.webview.postMessage(message)
  }

  readonly reveal: Panel<InMessage, OutMessage>['panel']['reveal'] = (...args) => {
    return this.panel.reveal(...args)
  }

  readonly onDidDispose: Panel<InMessage, OutMessage>['disposer']['onDidDispose'] = (...args) => {
    return this.disposer.onDidDispose(...args)
  }

  readonly dispose: Panel<InMessage, OutMessage>['disposer']['dispose'] = (...args) => {
    return this.disposer.dispose(...args)
  }

  readonly registerDisposable: Panel<InMessage, OutMessage>['disposer']['registerDisposable'] = (...args) => {
    return this.disposer.registerDisposable(...args)
  }

  readonly disposed: Panel<InMessage, OutMessage>['disposer']['disposed'] = (...args) => {
    return this.disposer.disposed(...args)
  }
}

interface Messageable<IncomingMessage, OutgoingMessage> {
  handleMessage: (message: IncomingMessage) => Promise<void>
  postMessage: (message: OutgoingMessage) => Promise<void>
}

interface Disposable {
  onDidDispose: (listener: () => void) => void
  dispose: () => void
  registerDisposable: <T extends vscode.Disposable>(value: T) => T
  disposed: () => boolean
}

export interface ExtensionHostContext {
  resourceRootDir: string
  client: LanguageClient
}

class Test extends Panel<void, void> {
  async handleMessage(message: void): Promise<void> {

  }

  constructor(context: ExtensionHostContext) {
    super(() => {
      const panel = vscode.window.createWebviewPanel(
        'test',
        'test',
        vscode.ViewColumn.One,
        {
          enableScripts: true
        }
      )
      return panel
    })
  }
}

export class PanelManager<U, V, T extends Panel<U, V>> {
  private panel: T | null = null

  constructor(
    private readonly context: ExtensionHostContext,
    private readonly PanelClass: new (context: ExtensionHostContext) => T
  ) {}

  newPanel(): void {
    if (this.panel != null) {
      this.panel.dispose()
    }
    this.panel = new this.PanelClass(this.context)
  }

  revealOrNew(): void {
    if (this.panel != null && !this.panel.disposed()) {
      this.panel.reveal()
      return
    }
    this.newPanel()
  }
}

// new PanelManager({} as unknown as ExtensionHostContext, Test)
