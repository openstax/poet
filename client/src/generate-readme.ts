import vscode from 'vscode'
import { requestGenerateReadme } from '../../common/src/requests'
import { ExtensionHostContext } from './panel'
import { expect, getRootPathUri } from './utils'

export function readmeGenerator(hostContext: ExtensionHostContext) {
  return async () => {
    await requestGenerateReadme(
      hostContext.client, {
        workspaceUri: expect(getRootPathUri(), 'Could not get root workspace uri').toString()
      }
    )
    void vscode.window.showInformationMessage('Done!', { title: 'Generate README' })
  }
}
