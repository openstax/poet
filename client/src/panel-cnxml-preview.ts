import vscode from 'vscode'
import fs from 'fs'
import path from 'path'
import { fixResourceReferences, fixCspSourceReferences, addBaseHref, expect, getRootPathUri, ensureCatchPromise, ensureCatch } from './utils'
import { PanelType } from './extension-types'
import { DOMParser, XMLSerializer } from 'xmldom'
import { ExtensionHostContext, Panel } from './panel'

// Line is one-indexed
export interface ScrollInEditorIncoming {
  type: 'scroll-in-editor'
  line: number
}

export interface DidReloadIncoming {
  type: 'did-reload'
}

export type PanelIncomingMessage = ScrollInEditorIncoming | DidReloadIncoming

export interface RefreshOutgoing {
  type: 'refresh'
  xml: string
  xsl: string
}

// Line is one-indexed
export interface ScrollToLineOutgoing {
  type: 'scroll-in-preview'
  line: number
}

export type PanelOutgoingMessage = (
  RefreshOutgoing
  | ScrollToLineOutgoing
)

const ELEMENT_NODE = 1
export const tagElementsWithLineNumbers = (doc: Document): void => {
  const root = doc.documentElement
  const stack: Element[] = [root]
  while (stack.length > 0) {
    const current = expect(stack.pop(), 'stack length is non-zero')
    current.setAttribute('data-line', (current as any).lineNumber)
    stack.push(...Array.from(current.childNodes).filter(node => node.nodeType === ELEMENT_NODE) as Element[])
  }
}

const initPanel = (context: ExtensionHostContext): vscode.WebviewPanel => {
  const editor = vscode.window.activeTextEditor
  const resourceColumn = editor?.viewColumn ?? vscode.ViewColumn.One
  const previewColumn = resourceColumn + 1
  const localResourceRoots = [vscode.Uri.file(context.resourceRootDir), expect(getRootPathUri(), 'workspace must be open to preview')]
  const panel = vscode.window.createWebviewPanel(
    PanelType.CNXML_PREVIEW,
    'CNXML Preview',
    previewColumn, {
      enableScripts: true,
      localResourceRoots,
      enableFindWidget: true
    }
  )
  panel.webview.html = rawTextHtml('Loading...')
  return panel
}

export const isXmlFile = (document: vscode.TextDocument): boolean => {
  return document.languageId === 'xml'
}

export const rawTextHtml = (text: string): string => {
  // Just to be safe...
  if (!/^[a-zA-Z0-9 ,.!]*$/.test(text)) {
    throw new Error('Must use simpler text for injection into HTML')
  }
  return `<html><body>${text}</body></html>`
}

export class CnxmlPreviewPanel extends Panel<PanelIncomingMessage, PanelOutgoingMessage> {
  private resourceBinding: vscode.Uri | null = null
  private webviewIsScrolling: boolean = false
  private resourceIsScrolling: boolean = false
  private xsl: string = ''
  private readonly _onDidChangeResourceBinding: vscode.EventEmitter<vscode.Uri | null>
  private readonly _onDidInnerPanelReload: vscode.EventEmitter<void>
  constructor(private readonly context: ExtensionHostContext) {
    super(initPanel(context))
    this._onDidChangeResourceBinding = new vscode.EventEmitter()
    this._onDidInnerPanelReload = new vscode.EventEmitter()
    void ensureCatchPromise(this.tryRebindToActiveResource(true))
    this.registerDisposable(vscode.window.onDidChangeActiveTextEditor((editor) => {
      void ensureCatchPromise(this.tryRebindToActiveResource(false))
    }))
    this.registerDisposable(this.context.events.onDidChangeWatchedFiles(ensureCatch(async () => {
      await this.tryRebindToActiveResource(false)
    })))
    this.registerDisposable(vscode.window.onDidChangeTextEditorVisibleRanges(event => {
      void ensureCatchPromise(this.tryScrollToRangeStartOfEditor(event.textEditor))
    }))
  }

  readonly onDidChangeResourceBinding: vscode.Event<vscode.Uri | null> = (...args) => {
    return this._onDidChangeResourceBinding.event(...args)
  }

  async handleMessage(message: PanelIncomingMessage): Promise<void> {
    if (message.type === 'scroll-in-editor') {
      for (const editor of vscode.window.visibleTextEditors) {
        if (!this.isPreviewOf(editor.document.uri)) {
          continue
        }
        if (this.resourceIsScrolling) {
          this.resourceIsScrolling = false
          return
        }
        this.webviewIsScrolling = true
        const line = message.line
        const sourceLine = Math.floor(line)
        const fraction = line - sourceLine
        const text = editor.document.lineAt(sourceLine).text
        const start = Math.floor(fraction * text.length)
        const range = new vscode.Range(sourceLine - 1, start, sourceLine, 0)
        const strategy = vscode.TextEditorRevealType.AtTop
        editor.revealRange(range, strategy)
      }
    } else {
      throw new Error(`Unexpected message: ${JSON.stringify(message)}`)
    }
  }

  private async tryScrollToRangeStartOfEditor(editor: vscode.TextEditor): Promise<void> {
    if (!this.isPreviewOf(editor.document.uri)) {
      return
    }
    if (this.webviewIsScrolling) {
      this.webviewIsScrolling = false
      return
    }
    this.resourceIsScrolling = true
    await this.scrollToRangeStartOfEditor(editor)
  }

  private async scrollToRangeStartOfEditor(editor: vscode.TextEditor): Promise<void> {
    const firstVisiblePosition = editor.visibleRanges[0].start
    const lineNumber = firstVisiblePosition.line
    const lineContent = editor.document.lineAt(lineNumber)
    const progress = firstVisiblePosition.character / (lineContent.text.length + 2)
    await this.postMessage({ type: 'scroll-in-preview', line: lineNumber + progress + 1 })
  }

  async tryRebindToActiveResource(force: boolean): Promise<void> {
    const activeCnxml = this.activeCnxmlUri()
    await this.tryRebindToResource(activeCnxml, force)
  }

  async tryRebindToResource(resource: vscode.Uri | null, force: boolean): Promise<void> {
    if (resource == null && !force) {
      return
    }
    const bindingChanged = resource?.fsPath !== this.resourceBinding?.fsPath
    await this.rebindToResource(resource)
    const activeEditor = vscode.window.activeTextEditor
    if (activeEditor != null) {
      await this.scrollToRangeStartOfEditor(activeEditor)
    }
    if (bindingChanged || force) {
      this._onDidChangeResourceBinding.fire(resource)
    }
  }

  private activeCnxmlUri(): vscode.Uri | null {
    const activeEditor = vscode.window.activeTextEditor
    if (activeEditor == null) {
      return null
    }
    const activeDocument = activeEditor.document
    if (!isXmlFile(activeDocument)) {
      return null
    }
    const activeUri = activeDocument.uri
    if (!(path.extname(activeUri.fsPath) === '.cnxml')) {
      return null
    }
    return activeUri
  }

  private async refreshContentsMessage(resource: vscode.Uri): Promise<PanelOutgoingMessage> {
    // TODO: Get resource contents from the language server?
    const contents = await fs.promises.readFile(resource.fsPath, { encoding: 'utf-8' })
    const doc = new DOMParser().parseFromString(contents)
    tagElementsWithLineNumbers(doc)
    const lineTaggedContents = new XMLSerializer().serializeToString(doc)
    // Load XSL if we haven't already
    if (this.xsl === '') {
      this.xsl = await fs.promises.readFile(
        path.join(this.context.resourceRootDir, 'cnxml-to-html5.xsl'),
        'utf-8'
      )
    }
    return { type: 'refresh', xml: lineTaggedContents, xsl: this.xsl }
  }

  isPreviewOf(resource: vscode.Uri | null): boolean {
    return resource?.fsPath === this.resourceBinding?.fsPath
  }

  private async rebindToResource(resource: vscode.Uri | null): Promise<void> {
    const oldBinding = this.resourceBinding
    this.resourceBinding = resource
    if (this.resourceBinding == null) {
      this.panel.webview.html = rawTextHtml('No resource available to preview')
      return
    }
    const message = await this.refreshContentsMessage(this.resourceBinding)
    if (oldBinding == null) {
      const html = await this.reboundWebviewHtmlForResource(this.resourceBinding, [message])
      this.panel.webview.html = html
    } else {
      void this.postMessage(message)
    }
  }

  private async reboundWebviewHtmlForResource(resource: vscode.Uri, messages: PanelOutgoingMessage[] = []): Promise<string> {
    let html = await fs.promises.readFile(path.join(this.context.resourceRootDir, 'cnxml-preview.html'), 'utf-8')
    html = this.injectEnsuredMessages(html, messages)
    html = addBaseHref(this.panel.webview, resource, html)
    html = fixResourceReferences(this.panel.webview, html, this.context.resourceRootDir)
    html = fixCspSourceReferences(this.panel.webview, html)
    return html
  }
}
