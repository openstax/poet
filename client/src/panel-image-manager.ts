import vscode from 'vscode'
import fs from 'fs'
import path from 'path'
import { fixResourceReferences, fixCspSourceReferences, getRootPathUri, expect } from './utils'
import { PanelType } from './extension-types'
import { type ExtensionHostContext, Panel } from './panel'

export interface PanelIncomingMessage {
  mediaUploads: Array<{ mediaName: string, data: string }>
}

const initPanel = (context: ExtensionHostContext): vscode.WebviewPanel => {
  const localResourceRoots = [vscode.Uri.file(context.resourceRootDir)]
  const workspaceRoot = getRootPathUri()
  /* istanbul ignore else */
  if (workspaceRoot != null) {
    localResourceRoots.push(workspaceRoot)
  }
  const panel = vscode.window.createWebviewPanel(
    PanelType.IMAGE_MANAGER,
    'Image Manager',
    vscode.ViewColumn.One,
    {
      enableScripts: true,
      localResourceRoots
    }
  )

  let html = fs.readFileSync(path.join(context.resourceRootDir, 'image-upload.html'), 'utf-8')
  html = fixResourceReferences(panel.webview, html, context.resourceRootDir)
  html = fixCspSourceReferences(panel.webview, html)
  panel.webview.html = html
  return panel
}

export class ImageManagerPanel extends Panel<PanelIncomingMessage, void, void> {
  constructor(private readonly context: ExtensionHostContext) {
    super(initPanel(context))
  }

  /* istanbul ignore next */
  protected getState() { throw new Error('BUG: Not implemented') }

  async handleMessage(message: PanelIncomingMessage): Promise<void> {
    const { mediaUploads } = message
    const uri = expect(getRootPathUri(), 'panel must act on the root workspace')
    for (const upload of mediaUploads) {
      const { mediaName, data } = upload
      const newFileUri = uri.with({ path: path.join(uri.path, 'media', mediaName) })
      try {
        await vscode.workspace.fs.stat(newFileUri)
        // FIXME: File exists already, do nothing for now. Maybe we should confirm the action?
        return
      } catch (err) {
        /* istanbul ignore else */
        if (err instanceof vscode.FileSystemError && err.name.includes('EntryNotFound')) {
          const content = Buffer.from(data.split(',')[1], 'base64')
          await vscode.workspace.fs.writeFile(newFileUri, content)
        }
      }
    }
  }
}
