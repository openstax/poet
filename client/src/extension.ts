import path from 'path'
import fs from 'fs'
import vscode from 'vscode'
import { LanguageClient } from 'vscode-languageclient/node'
import { pushContent, validateContent, setDefaultGitConfig } from './push-content'
import { TocEditorPanel } from './panel-toc-editor'
import { CnxmlPreviewPanel } from './panel-cnxml-preview'
import { expect, ensureCatch, ensureCatchPromise, launchLanguageServer, populateXsdSchemaFiles, getRootPathUri, configureWorkspaceSettings } from './utils'
import { OpenstaxCommand } from './extension-types'
import { ExtensionHostContext, Panel, PanelManager } from './panel'
import { ImageManagerPanel } from './panel-image-manager'
import { toggleTocTreesFilteringHandler } from './toc-trees-provider'
import { BookOrTocNode, TocsTreeProvider } from './book-tocs'
import { BooksAndOrphans, EMPTY_BOOKS_AND_ORPHANS, ExtensionServerNotification } from '../../common/src/requests'
import { writeReadmeForWorkspace } from './readme-generator'

let tocTreesView: vscode.TreeView<BookOrTocNode>
let tocTreesProvider: TocsTreeProvider
let client: LanguageClient
const onDidChangeWatchedFilesEmitter: vscode.EventEmitter<void> = new vscode.EventEmitter()
const onDidChangeWatchedFiles = onDidChangeWatchedFilesEmitter.event

let resourceRootDir = path.join(__dirname, 'static-resources')
let languageServerLauncher = launchLanguageServer
// setters for testing
export function setResourceRootDir(d: string) {
  resourceRootDir = d
}
export function setLanguageServerLauncher(l: typeof languageServerLauncher) {
  languageServerLauncher = l
}
export const forwardOnDidChangeWorkspaceFolders = (clientInner: LanguageClient) => async (event: vscode.WorkspaceFoldersChangeEvent) => {
  await clientInner.sendRequest('onDidChangeWorkspaceFolders', event)
}

type ExtensionExports = { [key in OpenstaxCommand]: PanelManager<Panel<unknown, unknown, unknown>> }
export async function activate(context: vscode.ExtensionContext): Promise<ExtensionExports> {
  // detect Theia. Alert the user if they are running Theia
  /* istanbul ignore next */
  expect(process.env.GITPOD_HOST != null && process.env.EDITOR?.includes('code') === false ? undefined : true, 'You seem to be running the Theia editor. Change your Settings in your profile')

  await configureWorkspaceSettings()

  client = languageServerLauncher(context)
  // Start the client. This will also launch the server
  client.start()

  // If this is not a book repo then don't bother writing the XSD files
  const workspaceRoot = getRootPathUri()
  /* The following istanbul comments are a hack because the coverage seems to misrepresent them as uncovered */
  /* istanbul ignore next */
  if (workspaceRoot !== null && fs.existsSync(path.join(workspaceRoot.fsPath, 'META-INF/books.xml'))) {
    await populateXsdSchemaFiles(resourceRootDir)
  }

  /* istanbul ignore next */
  await client.onReady()
  /* istanbul ignore next */
  const extExports = doRest(client)
  /* istanbul ignore next */
  return extExports
}

export async function deactivate(): Promise<void> {
  await expect(client, 'Expected client to have been activated').stop()
}

function createHostContext(client: LanguageClient): ExtensionHostContext {
  return {
    resourceRootDir,
    client, // FIXME: only pass in client.sendRequest, so as to disallow anything from calling onRequest
    events: {
      onDidChangeWatchedFiles
    },
    bookTocs: EMPTY_BOOKS_AND_ORPHANS
  }
}

function createExports(tocPanelManager: PanelManager<TocEditorPanel>, cnxmlPreviewPanelManager: PanelManager<CnxmlPreviewPanel>, imageManagerPanelManager: PanelManager<ImageManagerPanel>): ExtensionExports {
  return {
    [OpenstaxCommand.SHOW_TOC_EDITOR]: tocPanelManager,
    [OpenstaxCommand.SHOW_CNXML_PREVIEW]: cnxmlPreviewPanelManager,
    [OpenstaxCommand.SHOW_IMAGE_MANAGER]: imageManagerPanelManager
  }
}

function doRest(client: LanguageClient): ExtensionExports {
  const hostContext = createHostContext(client)
  const tocPanelManager = new PanelManager(hostContext, TocEditorPanel)
  const cnxmlPreviewPanelManager = new PanelManager(hostContext, CnxmlPreviewPanel)
  const imageManagerPanelManager = new PanelManager(hostContext, ImageManagerPanel)

  tocTreesProvider = new TocsTreeProvider()
  client.onNotification(ExtensionServerNotification.BookTocs, (params: BooksAndOrphans) => {
    hostContext.bookTocs = params // When a panel opens, make sure it has the latest bookTocs
    tocTreesProvider.update(params.books)
    /* istanbul ignore next */
    void tocPanelManager.panel()?.update(params)
  })

  vscode.workspace.onDidChangeWorkspaceFolders(ensureCatch(forwardOnDidChangeWorkspaceFolders(client)))
  vscode.commands.registerCommand(OpenstaxCommand.SHOW_TOC_EDITOR, tocPanelManager.revealOrNew.bind(tocPanelManager))
  vscode.commands.registerCommand(OpenstaxCommand.SHOW_IMAGE_MANAGER, imageManagerPanelManager.revealOrNew.bind(imageManagerPanelManager))
  vscode.commands.registerCommand(OpenstaxCommand.SHOW_CNXML_PREVIEW, cnxmlPreviewPanelManager.revealOrNew.bind(cnxmlPreviewPanelManager))
  vscode.commands.registerCommand('openstax.pushContent', ensureCatch(pushContent(hostContext)))
  vscode.commands.registerCommand('openstax.generateReadme', ensureCatch(writeReadmeForWorkspace))
  tocTreesView = vscode.window.createTreeView('tocTrees', { treeDataProvider: tocTreesProvider, showCollapseAll: true })
  vscode.commands.registerCommand('openstax.toggleTocTreesFiltering', ensureCatch(toggleTocTreesFilteringHandler(tocTreesView, tocTreesProvider)))
  vscode.commands.registerCommand('openstax.validateContent', ensureCatch(validateContent))
  void ensureCatchPromise(setDefaultGitConfig())

  // It is a logic error for anything else to listen to this event from the client.
  // It is only allowed a single handler, from what we can tell
  client.onRequest('onDidChangeWatchedFiles', () => { onDidChangeWatchedFilesEmitter.fire() })

  return createExports(tocPanelManager, cnxmlPreviewPanelManager, imageManagerPanelManager)
}
