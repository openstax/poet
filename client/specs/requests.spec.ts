import Sinon from 'sinon'
import { requestEnsureIds } from '../../common/src/requests'
import expect from 'expect'

describe('ensureIds', () => {
  const sinon = Sinon.createSandbox()
  afterEach(() => sinon.restore())

  it('runs and yields a response', async () => {
    const client = {
      sendRequest: sinon.stub()
    }
    await requestEnsureIds(client, { workspaceUri: '/fake/workspace' })
    expect(client.sendRequest.firstCall.args).toMatchSnapshot()
  })
})
