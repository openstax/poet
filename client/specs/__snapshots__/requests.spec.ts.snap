// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureIds runs and yields a response 1`] = `
Array [
  "bundle-ensure-ids",
  Object {
    "workspaceUri": "/fake/workspace",
  },
]
`;
